src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java

```diff
@@ -1144,6 +1150,33 @@ public boolean isKnownParticipant( UUID studyId, String participantId ) {
         return result;
     }
 
+    @Scheduled (fixedRate = 60000)
+    public void refreshUserAppsEntityIds() {
+        logger.info( "refreshing chronicle_user_apps entity ids" );
+
+        try {
+            ApiClient apiClient = prodApiClientCache.get( ApiClient.class );
+            DataApi dataApi = apiClient.getDataApi();
+            String jwtToken = auth0Client.getIdToken( username, password );
+
+            // load entities from chronicle_user_apps
+            Iterable<SetMultimap<FullQualifiedName, Object>> data = dataApi
```

Comment:
BEAUTIFUL

My only suggestion would be to cache the actual app name strings (that are used to generate the entity id) rather than the entityKeyIds themselves -- that way if a new app name comes in, an EKID is reserved, but then the DataApi write doesn't go through for whatever reason (chronicle server goes down, prod goes down, the write path is broken, etc) you won't be left with an empty entity that never gets populated.

This also lets you reduce the size of this call quite a bit -- DataApi lets you pass a filter on property types, so you can *only* load data from the property type used to generate the entity id (looks like `FULL_NAME_FQN` is the one).

Reply:
Actually, both the `FULL_NAME_FQN` and `FULL_NAME_FQN` are used to generate the entity id

Reply:
Maybe I'm misunderstanding what this method does? It looks like `FULL_NAME_FQN` is the only one that's passed through
```
private UUID reserveUserAppEntityKeyId(
            Map<UUID, Set<Object>> entityData,
            DataIntegrationApi dataIntegrationApi ) {

        return reserveEntityKeyId(
                entitySetIdMap.get( CHRONICLE_USER_APPS ),
                ImmutableList.of( propertyTypeIdsByFQN.get( FULL_NAME_FQN ) ),
                entityData,
                dataIntegrationApi
        );
    }
```

Reply:
Doesn't the EKID generated depend on the entityData? 
And the entity data has both FULL_NAME_FQN and TITLE_FQN

```
      Map<UUID, Set<Object>> userAppEntityData = new HashMap<>();
      userAppEntityData.put( propertyTypeIdsByFQN.get( FULL_NAME_FQN ), Sets.newHashSet( appPackageName ) );
      userAppEntityData.put( propertyTypeIdsByFQN.get( TITLE_FQN ), Sets.newHashSet( appName ) );
```

Reply:
It's a good idea to avoid having to load the entire entity set every refresh. Perhaps we can store a `Tuple` of `FULL_NAME_FQN` and `TITLE_FQN` that maps to the EKID?

Reply:
creating a map will take more space in memory and I don't understand how it would help avoid loading the entire entity set during refresh.

Reply:
@katherinebernstein even if we were to store the app name strings, we would still need an entity key id in order to write the subsequent associations.

Reply:
A few comments:

1) Your current implementation *does* only use `FULL_NAME_FQN` to derive the entityId. As you can see in `reserveUserAppEntityKeyId`, the only property type that's being passed to `reserveEntityKeyId` is `FULL_NAME_FQN` which means that's the only value that will be used in the default entity id generation. 

2) Because entityKeyIds are directly mapped by entityIds, and entityIds are directly mapped by specific entity values (i.e. FULL_NAME_FQN value), if two entities have the same FULL_NAME_FQN value they will necessarily have the same entityKeyId. This means you don't actually need to store entityKeyId -- you only need to store FULL_NAME_FQN value, and this way you guarantee you won't experience data loss which is currently very possible.

When you write a new entity, it's a two-step process. First, you reserve an entityKeyId for a designated entityId, which effectively creates an entity in the entity set (at this point, with no actual data attached to it). That empty entity will be returned by requests to load data from the entity set. Second, you write data to that entity. If that second step fails, given the current approach, **you will never populate that entity data**. This is why I'm suggesting storing the actual value of `FULL_NAME_FQN` (vs the entity key id) since that will safeguard against the failure case.

(Also if you really want to store the EKID you could theoretically keep a `Map<String, UUID>` from FULL_NAME_FQN value to entityKeyId, just to sanity check entityKeyIds as an extra check that the entityId mapping never changes (which would break a lot of things).

I think this is really really important -- **you are at risk of data loss if this is deployed currently**.

3) You don't need to worry about memory constraints since there's only like a couple thousand entities at most in this entity set, and the FULL_NAME_FQN strings are not very long.

4) You can avoid loading the entire entity set because DataApi exposes an endpoint which lets you specify which property types you want returned from an entity set, and will *only* return those property values. So you could call `DataApi.loadSelectedEntitySetData` and specify a set containing *only* FULL_NAME_FQN so that's the only data that comes back (and not all the other property types on that entity set that aren't used to derive the entityId since you don't care about those)

Reply:
Thanks for the extensive explanation! Totally makes sense. Fixed in https://github.com/openlattice/chronicle-server/pull/129